Feature/cms 1091 ability to move entries to a new section

i-just opened the pull request:
### Description
Adds element index action to move entries between sections.

- select entries to be moved and trigger the “Move to” action
<img width="1014" alt="Screenshot 2024-03-06 at 12 06 58" src="https://github.com/craftcms/cms/assets/4500340/c141da80-9d24-4e49-b4ff-64397df950aa">

- see a modal with a list of compatible sections (sections that have the same entry type(s) as the entries we want to move)
<img width="1016" alt="Screenshot 2024-03-06 at 12 07 10" src="https://github.com/craftcms/cms/assets/4500340/c6fb1532-45b8-47e1-9f17-dfbc4b0d336d">

- choose a section to move to
- entries are moved to the new section
<img width="374" alt="Screenshot 2024-03-06 at 12 07 17" src="https://github.com/craftcms/cms/assets/4500340/c2cd65c0-6ec3-41ff-9d75-2f7ebacc9b74">

**Notes:**
- it’s not possible to move to a single section
- for structures, entries are moved to the end of the new structure as root elements
- if the section entries are moved to has a lower `maxAuthors` limit, no data is changed/lost for the entries, but the next time they’re saved, a validation error will be shown
- when an entry is moved, its revisions and drafts are moved too; if their entry type isn’t allowed by the section we’re moving to, you will still be able to see the draft/revision but not apply/revert to it (the same applies to revisions and draft for entry types that were removed from the section); thanks @brianjhanson for adjusting the awesome `craft-tooltip` to work with disabled elements!
<img width="1626" alt="Screenshot 2024-03-11 at 09 59 42" src="https://github.com/craftcms/cms/assets/4500340/f412e7db-0c76-489d-92c7-de3e6a2a1035">
<img width="1623" alt="Screenshot 2024-03-11 at 10 00 13" src="https://github.com/craftcms/cms/assets/4500340/a6fdbbb4-d172-40d4-aa2b-84958bbeeb79">



### Related issues
[cms-1091](https://linear.app/craftcms/issue/CMS-1091/ability-to-move-entries-to-a-new-section)

i-just committed:
move entries to section element action (WIP)

i-just committed:
remove console logs

i-just committed:
make it work via keyboard

i-just committed:
Merge branch '5.0' into feature/cms-1091-ability-to-move-entries-to-a-new-section

i-just committed:
and now do it properly

i-just committed:
better messages

i-just committed:
quietly deal with authors incompatibilities

i-just committed:
permissions!

i-just committed:
Merge branch '5.0' into feature/cms-1091-ability-to-move-entries-to-a-new-section

i-just committed:
Merge branch '5.0' into feature/cms-1091-ability-to-move-entries-to-a-new-section

i-just committed:
allow move when too many authors

i-just committed:
move drafts & revisions too

i-just committed:
safeguard applying drafts & reverting to revisions

brianjhanson committed:
Update tooltip to work on disabled elements

i-just committed:
Merge branch '5.0' into feature/cms-1091-ability-to-move-entries-to-a-new-section

i-just committed:
Merge pull request #14561 from craftcms/feature/tooltip-on-diabled-elements

Update tooltip to work on disabled elements

i-just committed:
Merge branch 'feature/cms-1091-ability-to-move-entries-to-a-new-section' of https://github.com/craftcms/cms into feature/cms-1091-ability-to-move-entries-to-a-new-section

i-just committed:
tooltip for disabled revert to revision

i-just committed:
compiled assets

brandonkelly committed:
Merge branch '5.0' into feature/cms-1091-ability-to-move-entries-to-a-new-section

# Conflicts:
#	src/web/assets/cp/dist/cp.js
#	src/web/assets/cp/dist/cp.js.map
#	src/web/assets/cp/dist/css/cp.css
#	src/web/assets/cp/dist/css/cp.css.map

brandonkelly commented:
Rather than showing the “Couldn’t find any sections that all selected elements could be moved to.” message in the modal, ideally the “Move to” action is just grayed out ahead of time if there aren’t any other sections with the same entry type.

This should be doable via the `validateSelection()` method on the action trigger JS.

gcamacho079 reviewed (changes requested):
Just finished a quick scan of the modal contents:
- [x] The section buttons should have an `aria-pressed` value that toggles between true and false (see [internal documentation re: exclusive button groups](https://www.notion.so/pixelandtonic/exclusive-button-group-ed2321f5ffe34bbea5e42e2a93ac3624?pvs=4))
- [x] The Move button's `aria-disabled` state should be toggled when a section is selected
- [ ] The focus indicator for the move icon is under 3:1 contrast ratio against the modal footer background. This is a larger global issue caused by the disabled element opacity.
- [x] The focus indicator for the button is difficult to see since it doesn’t contrast against the dark background. This could benefit from a two-toned focus indicator, like the one used in the global skip link, with a white outline adjacent to the button background.
- [x] Loading spinner that appears in modal content body should have an equivalent status message that is read for screen reader users.

Note that some of these (like the disabled button focus outline) may be larger/more global issues than the scope of this PR.

i-just committed:
bug fix!

i-just committed:
Merge branch '5.0' into feature/cms-1091-ability-to-move-entries-to-a-new-section

i-just committed:
aria attrs

i-just committed:
lighter focus ring for dark elements

i-just committed:
check element is movable ahead of time

i-just committed:
Merge branch '5.0' into feature/cms-1091-ability-to-move-entries-to-a-new-section

i-just committed:
allow drafts too

i-just committed:
bug fix for nested elements

i-just committed:
reorganise js a bit

i-just committed:
maybe hooked up status messages

i-just committed:
temp

i-just committed:
Revert "temp"

This reverts commit a354d1834242f529becd5aa01e06a5e42b7d9ec0.

i-just committed:
keep the ability to add data attr to the button anyway

i-just committed:
add live region to modals & use in entry mover

i-just committed:
focus ring adjustments

i-just commented:
@gcamacho079, as per our chats, all items except the disabled button focus ring are actioned.

To get the last item from the list to work, I added the live region container and `updateLiveRegion` method to all modals (via `Garnish.Modal`) and started using it in my `EntryMover`, which means other modals could start taking advantage of it.

i-just committed:
Merge branch '5.0' into feature/cms-1091-ability-to-move-entries-to-a-new-section

brianjhanson committed:
Merge branch '5.0' of github.com:craftcms/cms into feature/cms-1091-ability-to-move-entries-to-a-new-section

brianjhanson committed:
Update tooltip labels

brianjhanson committed:
Fix tooltips for sidebar jobs

brandonkelly committed:
Merge branch '5.0' into feature/cms-1091-ability-to-move-entries-to-a-new-section

brandonkelly committed:
build

brandonkelly committed:
"Move to…"

brandonkelly committed:
Message tweaks

brandonkelly committed:
Merge branch '5.1' into feature/cms-1091-ability-to-move-entries-to-a-new-section

brandonkelly commented on src/controllers/ElementsController.php:
```diff
@@ -884,25 +884,52 @@ private function _additionalButtons(
-            $components[] = Html::button(Craft::t('app', 'Apply draft'), [
-                'class' => ['btn', 'secondary', 'formsubmit'],
+            /** @phpstan-ignore-next-line */
+            $disabled = $canonical->hasMethod('isEntryTypeCompatible') && !$element->isEntryTypeCompatible();
```
ElementsController shouldn’t be giving any special treatment to entries, so we’ll need to rethink how we do this.

i-just replied:
ok; addressed here: https://github.com/craftcms/cms/pull/14541/commits/f86e782c22369d5dc5c3dbe61a076855c50c63bd

brandonkelly commented on src/base/ElementInterface.php:
```diff
@@ -881,6 +881,15 @@ public function canDelete(User $user): bool;
+     * @return bool
+     * @since 5.0.0
+     */
+    public function canMove(User $user): bool;
```
It seems a little weird to introduce the concept of “moving” to all element types, when entries are the only one that would actually know what to do with it.

i-just replied:
As per our chat, this might end up being a way to go.
But, for now, I have reverted those changes and used a different approach where I added the ability to further customise element chip & card attributes, and I’m adding `data-movable` via those.

i-just committed:
different way of showing incompatible entry type tooltip

i-just committed:
keep moving within Entry (not Element)

i-just committed:
ability to add additional chip/card attrs from element type

brandonkelly committed:
Merge branch '5.x' into feature/cms-1091-ability-to-move-entries-to-a-new-section

# Conflicts:
#	.github/workflows/ci.yml
#	src/elements/Entry.php
#	src/web/assets/cp/dist/cp.js
#	src/web/assets/cp/dist/cp.js.map
#	src/web/assets/cp/dist/css/cp.css
#	src/web/assets/cp/dist/css/cp.css.map
#	src/web/assets/cp/src/js/CraftElementLabel.js
#	src/web/assets/garnish/dist/garnish.js
#	src/web/assets/garnish/dist/garnish.js.map

brandonkelly committed:
Merge branch '5.1' into feature/cms-1091-ability-to-move-entries-to-a-new-section

# Conflicts:
#	src/web/assets/cp/dist/cp.js
#	src/web/assets/cp/dist/cp.js.map
#	src/web/assets/cp/dist/css/cp.css
#	src/web/assets/cp/dist/css/cp.css.map

brandonkelly committed:
Set data-movable via htmlAttributes()

brandonkelly committed:
Merge branch '5.1' into feature/cms-1091-ability-to-move-entries-to-a-new-section

brandonkelly committed:
Cleanup

- Made Entry::canMove() private
- Moved moveEntryToSection() to the Entries service

brianjhanson committed:
Merge branch '5.3' of github.com:craftcms/cms into feature/cms-1091-ability-to-move-entries-to-a-new-section

brandonkelly committed:
Update `@since` tags

brandonkelly committed:
Show selectable section chips in the Move Entry modal

gcamacho079 commented:
A few accessibility improvements based on automated and manual testing:

- [x] The `checkbox` role doesn’t communicate the correct interaction pattern. Radio inputs would work better since the selections are exclusive.
  - [x] Radio inputs should be contained in a `fieldset` with a `legend`/`aria-labelledby` that references the title of the modal
  - [x] List markup should be removed as radio group markup should provide the necessary relationship information
- [x] The accessible name of the inputs should match the visible name (i.e., the section name). Right now, the label ID that’s referenced via `aria-labelledby` does not exist, so the name is coming from the input `title`.
- [x] Focus is dropped when an entry is moved. In these instances, focus should probably be moved to the `.elements` container. Ensure programmatic focus can be set by adding `tabindex="-1" to the container.

brandonkelly committed:
Merge branch '5.3' into feature/cms-1091-ability-to-move-entries-to-a-new-section

brandonkelly committed:
A11y improvements

brandonkelly committed:
<fieldset> + move the spinner to the button

Commit:
Fix id attribute for heading, remove radiogroup role from fieldset

brandonkelly committed:
Merge branch '5.3' into feature/cms-1091-ability-to-move-entries-to-a-new-section

brandonkelly committed:
Cleanup

brandonkelly committed:
Release notes
